### DISCUSSION TITLE
AS3 when an interface method return an interface

### DISCUSSION THREAD

**Discussion by ylazy** [airsdk/Adobe-Runtime-Support#3486] [Q&A] [answered]
Hi!

Please read this example:

class `Set` implements `IEnumerableObject`
class `Map` implements `IEnumerableObject`

interface `IEnumerableItem`
```as
function get source():IEnumerableObject
function set source(value:IEnumerableObject):void
```

class `SetItem` implement `IEnumerableItem`
```as
...
public function get source():Set // Error: interface method source in interface IEnumerableItem is implemented with an incompatible signature in class SetItem (1144)
public function set source(value:Set):void // Error: interface method source in interface IEnumerableItem is implemented with an incompatible signature in class SetItem (1144)
...
```

class `MapItem` implement `IEnumerableItem`
```as
...
public function get source():Map // Error: interface method source in interface IEnumerableItem is implemented with an incompatible signature in class MapItem (1144)
public function set source(value:Map):void // Error: interface method source in interface IEnumerableItem is implemented with an incompatible signature in class MapItem (1144)
...
```

`Set` and `Map` are `IEnumerableObject`, so why I got these errors?
Any ideas?

Thanks!

**Comment by ajwfrost** [comment, marked as answer]
You've defined `IEnumerableItem` as being able to set/get an `IEnumerableObject` using the `source` property.
But in the `SetItem` class, you're trying to define a propery `source` that only sets/gets a `Set` object... so that is different from the interface definition.

`Set` and `Map` might implement `IEnumerableObject` but they're not the same - as you've written it, you couldn't set a `Map` object onto a `SetItem` object. But with the `IEnumerableItem` interface, it says you should be able to set anything that's an `IEnumerableObject` - you can't then restrict it in this way.

So - if you implement an interface, you must follow the exact function declaration.

If you want to restrict what you set within each class's `source` property, you'll have to do that at runtime I think. The only other way I can think of doing this would be if we could implement 'generics'/templates.. but there's no compiler support for this I'm afraid.

**Reply by ylazy** [reply]
I got it. Thank you so much!
